### PR TITLE
[Merged by Bors] - optimize ballot eligibility validation

### LINKED ISSUE #4883: miner/oracle: optimize eligibility validation
the code below should be executed only for reference ballot. the sync is incredibly slow with that loop

image

also for current epoch we should get this data from tortoise rather then lru. lru is bad here, because if set to 10_000 and in fact there are 11_000 in active set then this cache will be threshed. tortoise on the other hand needs to store this data regardless, and it stores it in a compact form.

### PR DESCRIPTION
**Pull request by dshulyak**
closes: https://github.com/spacemeshos/go-spacemesh/issues/4883
related: https://github.com/spacemeshos/go-spacemesh/issues/4841

active set is validated only for reference ballot in current epoch, for all other ballots eligibilities validated against declared slots in the reference ballot. beside making validation more efficient, it will allow to prune active set data in future changes. reference ballots is still loading data from cached datastore, i plan to remove eventually but not in this change.

additionally i rewrote all tests for eligibility validator to get to 100% coverage and make them less dependent on the implementation.

### CODE REVIEW

**Comment by dshulyak** [comment]
i synced node with this change couple of times. another related improvement https://github.com/spacemeshos/go-spacemesh/issues/4927

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

File: proposals/interface.go [hunk 1]
```diff
@@ -25,14 +25,14 @@ type ballotDecoder interface {
...
-}
-
 type nonceFetcher interface {
 	VRFNonce(types.NodeID, types.EpochID) (types.VRFPostIndex, error)
 }
 
+type vrfVerifier interface {
+	Verify(types.NodeID, []byte, types.VrfSignature) bool
+}
+
```
Comment (countvonzero) [inline proposals/interface.go]: seems a no-op

**Review by countvonzero — APPROVED** [review]

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors cancel

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors merge

**Pull request closed without merging**